Depositor bot guide (#455)
lidofinance/docs: docs/guides/depositor-bot.md

# Lido depositor bot

## Introduction

> Depositor bot is a part of [Deposit Security Module](link to dsm)..

Reviewer: ```suggestion
Depositor bot is a part of [Deposit Security Module](/contracts/deposit-security-module).
```

The Depositor Bot obtains signed deposit messages from Council Daemons. Once a sufficient number of messages is
collected to constitute a quorum, the bot proceeds to initiate a deposit into the designated staking module. This
deposit is executed using
the [depositBufferedEther](https://docs.lido.fi/contracts/deposit-security-module#depositbufferedether) function within
the [DepositSecurityModule](https://docs.lido.fi/contracts/deposit-security-module) smart contract.

## Requirements

### Hardware

- 1-core CPU
- 2GB RAM

> ### Nodes

Reviewer: Does it really need nodes, or would RPC services be enough? :thinking:

Author: Changed

doing is optimising gas spending. An example of this is fetching `GAS_FEE_PERCENTILE_DAYS_HISTORY_1` days of gas history
and checking `GAS_FEE_PERCENTILE_1` bot will send transactions only if current gas price is less or equals to the
percentile. Also `GAS_PRIORITY_FEE_PERCENTILE`, `MIN_PRIORITY_FEE`, `MAX_PRIORITY_FEE` variables are used to calculate
`maxFeePerGas` and `maxPriorityFeePerGas` transaction parameters. The formula is:

```
priority = min(max(
      GAS_PRIORITY_FEE_PERCENTILE reward percentile of fee history for the last block,
      MIN_PRIORITY_FEE,
    ),
    MAX_PRIORITY_FEE,
)

maxFeePerGas = baseFeePerGas * 2 + priority
maxPriorityFeePerGas = priority
```

### Envs

Required variables are(mainnet):

| Variable                          | Default                                    | Description                                                                                                              |
|-----------------------------------|--------------------------------------------|--------------------------------------------------------------------------------------------------------------------------|

> | WEB3_RPC_ENDPOINTS                | -                                          | List of rpc endpoints that will be used to send requests comma separated (`,`)                                           |

Reviewer: not clear, tbh, what are the endpoints — both EL and Gnosis or what is the expected format?

Author: Added in the description

| WALLET_PRIVATE_KEY                | -                                          | Account private key                                                                                                      |
| CREATE_TRANSACTIONS               | false                                      | If true then tx will be send to blockchain                                                                               |

> | LIDO_LOCATOR                      | 0xC1d0b3DE6792Bf6b4b37EccdcC24e45978Cfd2Eb | Lido Locator address. Mainnet by default. Other networks could be found [here](https://docs.lido.fi/deployed-contracts/) |

Reviewer: ```suggestion
| LIDO_LOCATOR                      | 0xC1d0b3DE6792Bf6b4b37EccdcC24e45978Cfd2Eb | Lido Locator address. Mainnet by default. Other networks can be found [here](https://docs.lido.fi/deployed-contracts/) |
```

    cd depositor-bot
    ```
2. Install requirements
    ```bash
    poetry install
    ```
3. Run depositor bot
    ```bash
    poetry run python src/depositor.py
    ```
4. Verify in logs that depositor bot is performing validations, you should see logs of a kind:
    ```
    {"name": "bots.depositor", "levelname": "INFO", "funcName": "execute", "lineno": 121, "module": "depositor", "pathname": "/app/src/bots/depositor.py", "timestamp": 1729511569, "msg": "Do deposit to module with id: 1."}
    {"name": "bots.depositor", "levelname": "INFO", "funcName": "_deposit_to_module", "lineno": 210, "module": "depositor", "pathname": "/app/src/bots/depositor.py", "timestamp": 1729511569, "msg": "Checks failed. Skip deposit."}
    {"name": "bots.depositor", "levelname": "INFO", "funcName": "_deposit_to_module", "lineno": 194, "module": "depositor", "pathname": "/app/src/bots/depositor.py", "timestamp": 1729511569, "msg": "Calculations deposit recommendations.", "value": false, "is_mellow": false}
    ```

If you are facing problems, check what environment variables depositor bot is using, find a log
line `"msg": "Bot env variables"`

### Docker

> Docker image could be found [here](https://docs.lido.fi/guides/tooling#depositor-bot).

Reviewer: ```suggestion
Docker image can be found [here](https://docs.lido.fi/guides/tooling#depositor-bot).
```

# Lido depositor bot

## Introduction

> Depositor bot is a part of Deposit Security Module.

Reviewer 2: ```suggestion
Depositor bot is a part of [Deposit Security Module](link to dsm).
```

Author: Done

> The Depositor Bot obtains signed deposit messages from Council Daemons. Once a sufficient number of messages is

Reviewer 2: Killer feature of deposit bot is gas price checking add some details how it works.
Also add to envs list gas price envs

Author: Added

Reviewer: +to @F4ever 
see no mentions of this as of now

Author: @TheDZhon please check `How to use` section

collected to constitute a quorum, the bot proceeds to initiate a deposit into the designated staking module. This

> deposit is executed using the depositBufferedEther function within the "DepositSecurityModule" smart contract.

Reviewer 2: link to the function spec

Author: Done

## Requirements

### Hardware

- 1-core CPU
- 2GB RAM

### Nodes

- Ethereum EL node

> - Onchain transport node

Reviewer 2: ```suggestion
- Onchain databus transport node
```

Reviewer 2: Add comment that currently it's gnosis chain

Author: Done

### Envs

> Required variables are:

Reviewer 2: Add note that defaults is correct for mainnet

Author: Done

    ```bash
    poetry install
    ```
3. Run depositor bot
    ```bash
    poetry run python src/depositor.py
    ```
4. Verify in logs that depositor bot is performing validations, you should see logs of a kind:
    ```
    {"name": "bots.depositor", "levelname": "INFO", "funcName": "execute", "lineno": 121, "module": "depositor", "pathname": "/app/src/bots/depositor.py", "timestamp": 1729511569, "msg": "Do deposit to module with id: 1."}
    {"name": "bots.depositor", "levelname": "INFO", "funcName": "_deposit_to_module", "lineno": 210, "module": "depositor", "pathname": "/app/src/bots/depositor.py", "timestamp": 1729511569, "msg": "Checks failed. Skip deposit."}
    {"name": "bots.depositor", "levelname": "INFO", "funcName": "_deposit_to_module", "lineno": 194, "module": "depositor", "pathname": "/app/src/bots/depositor.py", "timestamp": 1729511569, "msg": "Calculations deposit recommendations.", "value": false, "is_mellow": false}
    ```
  
  If you are facing problems, check what environment variables depositor bot is using, find a log line `"msg": "Bot env variables"`

### Docker

Docker image could be found [here](https://docs.lido.fi/guides/tooling#depositor-bot).

> ## Monitoring

Author: Added examples in https://github.com/lidofinance/depositor-bot/pull/276

> ## Monitoring

Author: Added link
